File: epochix-vscode/src/parsers/huggingface.ts
/**
 * TypeScript port of src/epochix/parsers/huggingface.py
 */
import type { Parser, ParserContext, RawMetric } from "./base";

const HF_DICT_LINE = /^\s*\{['"]loss['"].*\}/;

export class HuggingFaceParser implements Parser {
  readonly name = "huggingface";
  readonly priority = 80;

  sniff(sampleLines: readonly string[]): number {
    const hits = sampleLines.filter((l) => HF_DICT_LINE.test(l)).length;
    return Math.min((hits / Math.max(sampleLines.length, 1)) * 5, 0.93);
  }

  parseLine(line: string, ctx: ParserContext): RawMetric[] {
    const stripped = line.trim();
    if (!stripped.startsWith("{")) return [];

    // Normalize Python dict literals to valid JSON
    const normalized = stripped
      .replace(/'/g, '"')
      .replace(/True/g, "true")
      .replace(/False/g, "false");

    let data: Record<string, unknown>;
    try {
      data = JSON.parse(normalized) as Record<string, unknown>;
    } catch {
      return [];
    }

    const epoch = data["epoch"];
    if (epoch !== undefined && (typeof epoch === "number")) {
      ctx.currentEpoch = epoch;
      delete data["epoch"];
    }

    const metrics: RawMetric[] = [];
    for (const [key, val] of Object.entries(data)) {
      if (typeof val !== "number") continue;
      metrics.push({
        seq: ctx.seq,
        epoch: ctx.currentEpoch,
        step: ctx.currentStep,
        key,
        value: val,
        parserName: this.name,
        confidence: 0.91,
      });
    }
    return metrics;
  }
}

File: src/epochix/parsers/huggingface.py
from __future__ import annotations

import json
import re

from epochix.models import RawMetric
from epochix.parsers.base import ParserContext
from epochix.parsers.registry import register_parser

# HF Trainer logs a JSON-like dict per step/eval:
#   {'loss': 0.5123, 'learning_rate': 5e-05, 'epoch': 1.0}
#   {'eval_loss': 0.3421, 'eval_accuracy': 0.8765, 'epoch': 1.0}
_HF_DICT_LINE = re.compile(r"^\s*\{['\"]loss['\"].*\}")


@register_parser
class HFParser:
    name = "huggingface"
    priority = 80

    def sniff(self, sample_lines: list[str]) -> float:
        hits = sum(1 for line in sample_lines if _HF_DICT_LINE.match(line))
        return min(hits / max(len(sample_lines), 1) * 5, 0.93)

    def parse_line(self, line: str, ctx: ParserContext) -> list[RawMetric]:
        stripped = line.strip()
        if not stripped.startswith("{"):
            return []

        # Normalize Python dict literals to valid JSON
        normalized = stripped.replace("'", '"').replace("True", "true").replace("False", "false")
        try:
            data: dict[str, object] = json.loads(normalized)
        except json.JSONDecodeError:
            return []

        epoch = data.pop("epoch", None)
        if epoch is not None and isinstance(epoch, (int, float)):
            ctx.current_epoch = float(epoch)

        # `step` is training progress, not a metric. Without popping it, the HF
        # parser (which also matches Accelerate's `accelerator.print({...})`
        # dicts) emitted the step count as a bogus `custom` metric on the
        # dashboard and never set the step context.
        step = data.pop("step", None)
        if step is not None and isinstance(step, (int, float)):
            ctx.current_step = int(step)

        metrics: list[RawMetric] = []
        for key, val in data.items():
            if not isinstance(val, (int, float)):
                continue
            metrics.append(
                RawMetric(
                    seq=ctx.seq,
                    epoch=ctx.current_epoch,
                    step=ctx.current_step,
                    key=key,
                    value=float(val),
                    parser_name=self.name,
                    confidence=0.91,
                )
            )
        return metrics
